asn1/InterledgerTypes.asn

```diff
@@ -0,0 +1,165 @@
+-- })
+
+-- Amount ::= SEQUENCE {
+--  mantissa UInt64,
+--  exponent Int8
+-- }
+
+-- --------------------------------------------------------------------------
+
+-- Different ledgers may use different rounding parameters
+
+-- NumberFormat ::= CHOICE {
+    -- Most ledgers today are designed for humans and use base-10. However, it
+    -- may be desirable to create base 2 ledgers in the future for efficiency.
+    -- base2 [2] IMPLICIT Base2NumberFormat,
+    -- base10 [10] IMPLICIT Base10NumberFormat
+-- }
+
+-- Base2NumberFormat ::= SEQUENCE {
+    -- precision UInt8,
+
+    -- Providing a range of exponents allows us to describe both
+    -- fixed-point (min == max) and floating-point (min != max) type ledgers.
+    -- minExponent Int16,
+    -- maxExponent Int16
+-- }
+
+-- Base10NumberFormat ::= SEQUENCE {
+    -- precision UInt8,
+
+    -- Providing a range of exponents allows us to describe both
+    -- fixed-point (min == max) and floating-point (min != max) type ledgers.
+    -- minExponent Int16,
+    -- maxExponent Int16
+
+    -- Stepping is used in some assets some of the time. For instance, Swiss
+    -- Francs are sometimes rounded in CHF 0.05 increments. However, this is an
+    -- obscure enough use case that we don't feel it's necessary to support.
+    -- Note that stepping is only possible for base-10 ledgers, not base-2,
+    -- step UInt8 ( 1 | 2 | 5 )
+
+    -- Rounding could be specified per ledger, but we believe that in most cases
+    -- rounding is dictated by security and self-interest. For example, a
+    -- connector will always round in its favor to prevent its customers from
+    -- being able to extract money from it over time by tricking it to round in
+    -- their favor over and over.
+    -- roundingMode ENUMERATED {
+    --  up (0)
+    --  down (1)
+    --  ceiling (2)
+    --  floor (3)
+    --  halfUp (4)
+    --  halfDown (5)
+    --  halfEven (6)
+    -- }
+--}
+
+-- --------------------------------------------------------------------------
+
+-- GeneralizedTime is a standard ASN.1 time format based on ISO 8601. It was
+-- tempting to use the Unix epoch. Unfortunately, standard POSIX time is
+-- discontinuous. If we used POSIX time for expiries for example, a transfer
+-- might expire and then "un-expire" when a leap second occurs. Since
+-- Interledger protocols depend on causality, we cannot use POSIX time and
+-- since we do not wish to create our own time standard, we end up using ISO
+-- 8601.
+
+Timestamp ::= GeneralizedTime
+
+-- --------------------------------------------------------------------------
+
+-- Liquidity curves describe the relationship between input and output amount
+-- for a given path between a given pair of ledgers.
+--
+-- The curve is expressed as a series of points given as <x, y> coordinates.
+--
+-- Each x coordinate corresponds to an input amount, each y corresponds to an
+-- output amount. Each point must have larger x and y values than the previous
```

Comment:
Are you sure both x and y have to be larger? Can you not have a "curve" that plateaus in the middle, e.g.:

```
   y₃ ┨         ╭┄
      ┃        ╱
      ┃       ╱
y₁,y₂ ┨  ╭╌╌╌╯
      ┃ ╱
   y₀ ┨╯
      ┗━━┳━━━┳━━┳━━
     x₀ x₁   x₂ x₃
```

(in this case y₁=y₂ but x₂>x₁)